### ISSUE TITLE
Problem: historical data loss after node restart

### ISSUE THREAD

**Issue by 1143857824** [XRPLF/rippled#4262]
**Problem:** historical data loss after node restart
**Expected result:** it is hoped that the data after the time when the node is first built can be retained, and the previous data will not be lost even if it is restarted

**How to configure nodes to achieve this purpose？**

The following is our current node setup configuration, but there is this problem at present

> rippled.cfg
    
    [server]
    port_rpc_admin_local
    port_peer
    port_ws_admin_local
    
    [port_rpc_admin_local]
    port = 5005
    ip = 0.0.0.0
    admin = 0.0.0.0
    protocol = http
    
    [port_peer]
    port = 51235
    ip = 0.0.0.0
    #ip = ::
    protocol = peer
    
    [port_ws_admin_local]
    port = 6006
    ip = 0.0.0.0
    admin = 0.0.0.0
    protocol = ws
    
    [node_size]
    medium
    
    [node_db]
    type=NuDB
    path=/var/lib/rippled/db/nudb
    online_delete=81920000
    advisory_delete=0
    
    [database_path]
    /var/lib/rippled/db
    
    [debug_logfile]
    /var/log/rippled/debug.log
    
    [sntp_servers]
    time.windows.com
    time.apple.com
    time.nist.gov
    pool.ntp.org
    
    
    [validators_file]
    validators.txt
    
    [rpc_startup]
    { "command": "log_level", "severity": "warning" }
    
    [ssl_verify]
    1

**Comment by ximinez** [comment]
`rippled` does not instantly know what data is available on startup. It basically has to backfill in the same way it would if the data was not there, except that it is much faster, because it's available locally. Additionally, it will first work to retrieve any ledgers that were missed while the node was down _before_ it finds those ledgers locally. With an `online_delete` value that large, it will take a significant amount of time to do that work, but it will eventually get there.

tl;dr Give your node more time to "discover" that it has those ledgers after all.

**Comment by 1143857824** [comment]
@ximinez Thank you very much for being able to answer my question.
I don't understand it very well. When I start rippled for the first time, I can get the data after the height "70144088", but after restarting it once, I can only access the data after the height "72378219". is this normal?
Shouldn't the data after the "70144088" height be written to the hard disk from the first boot? Why can't the block be accessed after restarting?

**Comment by ximinez** [comment]
How are you trying to access the ledger? There are some ways of requesting the info which will go directly to the database, and others which require rippled to have processed back to the ledger.

For example, I ran these commands on my local node:
```
$ rippled -q json ledger '{"ledger_index": 73469653, "transactions": true }'
{
   "result" : {
      "ledger" : {
         "accepted" : true,
         "account_hash" : "1D43AFDF9A15FBC0BD589240F0487E89425ABA81E700B9505EF73FDFCC1FA552",
         "close_flags" : 0,
         "close_time" : 712945901,
         "close_time_human" : "2022-Aug-04 16:31:41.000000000 UTC",
         "close_time_resolution" : 10,
         "closed" : true,
         "hash" : "9DE6ECF7529B4DDB517A6D69D8994CE9B15C9DB1434E0203626180814A7E6873",
         "ledger_hash" : "9DE6ECF7529B4DDB517A6D69D8994CE9B15C9DB1434E0203626180814A7E6873",
         "ledger_index" : "73469653",
         "parent_close_time" : 712945900,
         "parent_hash" : "F1698F1D6B6049B91F7FA9E2923C2DE9B92553373BF3D9630322A2DCAAF3B792",
         "seqNum" : "73469653",
         "totalCoins" : "99989365827414354",
         "total_coins" : "99989365827414354",
         "transaction_hash" : "95C1BBCC30D58D089225BDC09A3CA1EC159C42DAE3620870C03D0F443CF5DAEE",
         "transactions" : [
            "0810CDF6AFE619DDFED1F272349C580B04D78446375AF587FD6169697DFFBE4B",
            "165E9838B3F93F7B8ABA5E9AD4A0916025262E54990CEC44169E9A5D14D2A284",
            "18A48FB68E35AF1AC6F6AD59DA8FDB2563538A8A724F02A1FAC655B91460C31A",
            "21D7B7DC1799F01BA9E7A7D205E2DCCB3FCDC3F8D2F43782900050919F87BBB0",
            "2E731D77A36CD7245680B4694C4C982FF5983000B0331A5A0731B65BD59B17B9",
            "48D834AE692F0F9B80E282E3933BD00BE9FFE9425D58D5087DB65AEABC39877B",
            "540237183D2FBB34AD42E7298975F135C639D619AAA78FAC7B4A560EFD77F2D4",
            "55C90CC7B321BD0855A2E431BAF4D03FBC5AC88E073BB35B5265898F4E5BAA28",
            "5D98D0CB6FAB57D5C954B7F8309F4849D6F03ED49C20D18DD71A564326DB0D2B",
            "6ADE854F15B5D3A9F981FBC42FD1C1F361585D48B69E2CA9A634B97945CB9480",
            "6BFDFF169479B135C14A07DD8DA7E59DA93FEB4637BDD54E6A147575BE64D389",
            "6FC2565E80200F5F1D700CF6787326FA5F0F02C4D9C82D33829F92E559D92E24",
            "7789B664C8BAC4BD952FD6CE22264F7AEB3DA5317D24BBDE0B27A652A5D2CE2F",
            "79E120DCACE148F38DA161F2B8B6F02CD27F51D8052B301B768EACDE29B10220",
            "813B00A348EADE502C86B0FB60E4AD66671997518CDCF0118594A31BDD0B89F2",
            "855DE537DA4D28950563544D2F94019B1BDE0EFD3F49B3A075F01456A386B372",
            "870F603273F8E7BEC54624C72124CD4931C4383807BC8810723F6E65BF76F55C",
            "8DEE15F70662D59779FBA88D790F417D465323452A2E9AB483CF9E6ECBB6BD3A",
            "9B4EB7C06F7D9D7B261DD1685EA9AD97ED6228B93EFEE1368D927C17FDF8CCF2",
            "A67AB4F51EA2086F048A3299CF519DEBD08A1CC35969AD34D572EDF78D9664A8",
            "ABF81A03F2BCBC680A8204A9E6C8367EAB5C4831475E2081FFA656D57EADAC61",
            "B6DFAE6B40BDBC10D9805C4CC5867D804C8AA997B96445E16834D31017E3BFB6",
            "C6BC6D6FB2F77F03F178200D483FD20CE1A7FDE87E97629040CA953782AC002D",
            "D2FCDC13529A8586B0313F78E6499DCA5D946236007811139170DF7AED866870",
            "D8131A724817F4FDE969B52A924B44E984D94B0E2433A7AFBE056203F1BE8F41",
            "DBBF033763C839A026A1C7D6AAEFD88DDA72D7DD5617D40D2B5CACEDD2212156",
            "DE1A53940910DD8D69DEA1B1713E872A1CC16E02883187CA083FB3750474AA23",
            "E3750EE40504D7BF7FDDB9CEE16C9688590D2EE6F8FA01CC7A712C06F799727D",
            "EAA2DAED39BA5E506F99F5B446ECEF721684BA48E0783FFBD1A3EE748F8AC0CE",
            "F021CC33512D68ABD8AAE0A4C1AF1BEAE0D0637B11B32FC76691BD346CF34D4C",
            "F0FF32B7A58AF5AE394B893AB07280997138A4C7C1D298C114D7B9FADF2CB988",
            "F7689F4868E97B0B775AFF3939F7EBE0D5CFD950649829EF4C2019115381A603"
         ]
      },
      "ledger_hash" : "9DE6ECF7529B4DDB517A6D69D8994CE9B15C9DB1434E0203626180814A7E6873",
      "ledger_index" : 73469653,
      "status" : "success",
      "validated" : true
   }
}

$ rippled -q json ledger '{"ledger_hash": "9DE6ECF7529B4DDB517A6D69D8994CE9B15C9DB1434E0203626180814A7E6873", "transactions": true }'
{
   "result" : {
      "ledger" : {
         "accepted" : true,
         "account_hash" : "1D43AFDF9A15FBC0BD589240F0487E89425ABA81E700B9505EF73FDFCC1FA552",
         "close_flags" : 0,
         "close_time" : 712945901,
         "close_time_human" : "2022-Aug-04 16:31:41.000000000 UTC",
         "close_time_resolution" : 10,
         "closed" : true,
         "hash" : "9DE6ECF7529B4DDB517A6D69D8994CE9B15C9DB1434E0203626180814A7E6873",
         "ledger_hash" : "9DE6ECF7529B4DDB517A6D69D8994CE9B15C9DB1434E0203626180814A7E6873",
         "ledger_index" : "73469653",
         "parent_close_time" : 712945900,
         "parent_hash" : "F1698F1D6B6049B91F7FA9E2923C2DE9B92553373BF3D9630322A2DCAAF3B792",
         "seqNum" : "73469653",
         "totalCoins" : "99989365827414354",
         "total_coins" : "99989365827414354",
         "transaction_hash" : "95C1BBCC30D58D089225BDC09A3CA1EC159C42DAE3620870C03D0F443CF5DAEE",
         "transactions" : [
            "0810CDF6AFE619DDFED1F272349C580B04D78446375AF587FD6169697DFFBE4B",
            "165E9838B3F93F7B8ABA5E9AD4A0916025262E54990CEC44169E9A5D14D2A284",
            "18A48FB68E35AF1AC6F6AD59DA8FDB2563538A8A724F02A1FAC655B91460C31A",
            "21D7B7DC1799F01BA9E7A7D205E2DCCB3FCDC3F8D2F43782900050919F87BBB0",
            "2E731D77A36CD7245680B4694C4C982FF5983000B0331A5A0731B65BD59B17B9",
            "48D834AE692F0F9B80E282E3933BD00BE9FFE9425D58D5087DB65AEABC39877B",
            "540237183D2FBB34AD42E7298975F135C639D619AAA78FAC7B4A560EFD77F2D4",
            "55C90CC7B321BD0855A2E431BAF4D03FBC5AC88E073BB35B5265898F4E5BAA28",
            "5D98D0CB6FAB57D5C954B7F8309F4849D6F03ED49C20D18DD71A564326DB0D2B",
            "6ADE854F15B5D3A9F981FBC42FD1C1F361585D48B69E2CA9A634B97945CB9480",
            "6BFDFF169479B135C14A07DD8DA7E59DA93FEB4637BDD54E6A147575BE64D389",
            "6FC2565E80200F5F1D700CF6787326FA5F0F02C4D9C82D33829F92E559D92E24",
            "7789B664C8BAC4BD952FD6CE22264F7AEB3DA5317D24BBDE0B27A652A5D2CE2F",
            "79E120DCACE148F38DA161F2B8B6F02CD27F51D8052B301B768EACDE29B10220",
            "813B00A348EADE502C86B0FB60E4AD66671997518CDCF0118594A31BDD0B89F2",
            "855DE537DA4D28950563544D2F94019B1BDE0EFD3F49B3A075F01456A386B372",
            "870F603273F8E7BEC54624C72124CD4931C4383807BC8810723F6E65BF76F55C",
            "8DEE15F70662D59779FBA88D790F417D465323452A2E9AB483CF9E6ECBB6BD3A",
            "9B4EB7C06F7D9D7B261DD1685EA9AD97ED6228B93EFEE1368D927C17FDF8CCF2",
            "A67AB4F51EA2086F048A3299CF519DEBD08A1CC35969AD34D572EDF78D9664A8",
            "ABF81A03F2BCBC680A8204A9E6C8367EAB5C4831475E2081FFA656D57EADAC61",
            "B6DFAE6B40BDBC10D9805C4CC5867D804C8AA997B96445E16834D31017E3BFB6",
            "C6BC6D6FB2F77F03F178200D483FD20CE1A7FDE87E97629040CA953782AC002D",
            "D2FCDC13529A8586B0313F78E6499DCA5D946236007811139170DF7AED866870",
            "D8131A724817F4FDE969B52A924B44E984D94B0E2433A7AFBE056203F1BE8F41",
            "DBBF033763C839A026A1C7D6AAEFD88DDA72D7DD5617D40D2B5CACEDD2212156",
            "DE1A53940910DD8D69DEA1B1713E872A1CC16E02883187CA083FB3750474AA23",
            "E3750EE40504D7BF7FDDB9CEE16C9688590D2EE6F8FA01CC7A712C06F799727D",
            "EAA2DAED39BA5E506F99F5B446ECEF721684BA48E0783FFBD1A3EE748F8AC0CE",
            "F021CC33512D68ABD8AAE0A4C1AF1BEAE0D0637B11B32FC76691BD346CF34D4C",
            "F0FF32B7A58AF5AE394B893AB07280997138A4C7C1D298C114D7B9FADF2CB988",
            "F7689F4868E97B0B775AFF3939F7EBE0D5CFD950649829EF4C2019115381A603"
         ]
      },
      "ledger_hash" : "9DE6ECF7529B4DDB517A6D69D8994CE9B15C9DB1434E0203626180814A7E6873",
      "ledger_index" : 73469653,
      "status" : "success",
      "validated" : true
   }
}
$ rippled -q tx EAA2DAED39BA5E506F99F5B446ECEF721684BA48E0783FFBD1A3EE748F8AC0CE
{
   "result" : {
      "Account" : "rh3VLyj1GbQjX7eA15BwUagEhSrPHmLkSR",
      "Fee" : "15",
      "Flags" : 0,
      "LastLedgerSequence" : 73469655,
      "OfferSequence" : 109151505,
      "Sequence" : 109151509,
      "SigningPubKey" : "022D40673B44C82DEE1DDB8B9BB53DCCE4F97B27404DB850F068DD91D685E337EA",
      "TakerGets" : {
         "currency" : "CNY",
         "issuer" : "rJ1adrpGS3xsnQMb9Cw54tWJVFPuSdZHK",
         "value" : "4941.195410686269"
      },
      "TakerPays" : "2094996294",
      "TransactionType" : "OfferCreate",
      "TxnSignature" : "30450221008B2DBF4D6E7A332C67FE2720A3D79F39ED022B07B17DB9B5A9E7FD43F85A8E510220720596CC25B916ACF60E4F01FE7D60A7B75F57EE52B224CFFE6F7F35FAA2F2A6",
      "date" : 712945901,
      "hash" : "EAA2DAED39BA5E506F99F5B446ECEF721684BA48E0783FFBD1A3EE748F8AC0CE",
      "inLedger" : 73469653,
      "ledger_index" : 73469653,
      "meta" : {
[...]
         "TransactionIndex" : 17,
         "TransactionResult" : "tesSUCCESS"
      },
      "status" : "success",
      "validated" : true
   }
}
```

Then I stopped and restarted the node. Immediately after starting, I ran the same commands:

```
$ rippled -q json ledger '{"ledger_index": 73469653, "transactions": true }'
{
   "result" : {
      "error" : "noNetwork",
      "error_code" : 17,
      "error_message" : "InsufficientNetworkMode",
      "request" : {
         "api_version" : 1,
         "command" : "ledger",
         "ledger_index" : 73469653,
         "method" : "ledger",
         "transactions" : true
      },
      "status" : "error"
   }
}
$ rippled -q json ledger '{"ledger_hash": "9DE6ECF7529B4DDB517A6D69D8994CE9B15C9DB1434E0203626180814A7E6873", "transactions": true }'
{
   "result" : {
      "ledger" : {
         "accepted" : true,
         "account_hash" : "1D43AFDF9A15FBC0BD589240F0487E89425ABA81E700B9505EF73FDFCC1FA552",
         "close_flags" : 0,
         "close_time" : 712945901,
         "close_time_human" : "2022-Aug-04 16:31:41.000000000 UTC",
         "close_time_resolution" : 10,
         "closed" : true,
         "hash" : "9DE6ECF7529B4DDB517A6D69D8994CE9B15C9DB1434E0203626180814A7E6873",
         "ledger_hash" : "9DE6ECF7529B4DDB517A6D69D8994CE9B15C9DB1434E0203626180814A7E6873",
         "ledger_index" : "73469653",
         "parent_close_time" : 712945900,
         "parent_hash" : "F1698F1D6B6049B91F7FA9E2923C2DE9B92553373BF3D9630322A2DCAAF3B792",
         "seqNum" : "73469653",
         "totalCoins" : "99989365827414354",
         "total_coins" : "99989365827414354",
         "transaction_hash" : "95C1BBCC30D58D089225BDC09A3CA1EC159C42DAE3620870C03D0F443CF5DAEE",
         "transactions" : [
            "0810CDF6AFE619DDFED1F272349C580B04D78446375AF587FD6169697DFFBE4B",
            "165E9838B3F93F7B8ABA5E9AD4A0916025262E54990CEC44169E9A5D14D2A284",
            "18A48FB68E35AF1AC6F6AD59DA8FDB2563538A8A724F02A1FAC655B91460C31A",
            "21D7B7DC1799F01BA9E7A7D205E2DCCB3FCDC3F8D2F43782900050919F87BBB0",
            "2E731D77A36CD7245680B4694C4C982FF5983000B0331A5A0731B65BD59B17B9",
            "48D834AE692F0F9B80E282E3933BD00BE9FFE9425D58D5087DB65AEABC39877B",
            "540237183D2FBB34AD42E7298975F135C639D619AAA78FAC7B4A560EFD77F2D4",
            "55C90CC7B321BD0855A2E431BAF4D03FBC5AC88E073BB35B5265898F4E5BAA28",
            "5D98D0CB6FAB57D5C954B7F8309F4849D6F03ED49C20D18DD71A564326DB0D2B",
            "6ADE854F15B5D3A9F981FBC42FD1C1F361585D48B69E2CA9A634B97945CB9480",
            "6BFDFF169479B135C14A07DD8DA7E59DA93FEB4637BDD54E6A147575BE64D389",
            "6FC2565E80200F5F1D700CF6787326FA5F0F02C4D9C82D33829F92E559D92E24",
            "7789B664C8BAC4BD952FD6CE22264F7AEB3DA5317D24BBDE0B27A652A5D2CE2F",
            "79E120DCACE148F38DA161F2B8B6F02CD27F51D8052B301B768EACDE29B10220",
            "813B00A348EADE502C86B0FB60E4AD66671997518CDCF0118594A31BDD0B89F2",
            "855DE537DA4D28950563544D2F94019B1BDE0EFD3F49B3A075F01456A386B372",
            "870F603273F8E7BEC54624C72124CD4931C4383807BC8810723F6E65BF76F55C",
            "8DEE15F70662D59779FBA88D790F417D465323452A2E9AB483CF9E6ECBB6BD3A",
            "9B4EB7C06F7D9D7B261DD1685EA9AD97ED6228B93EFEE1368D927C17FDF8CCF2",
            "A67AB4F51EA2086F048A3299CF519DEBD08A1CC35969AD34D572EDF78D9664A8",
            "ABF81A03F2BCBC680A8204A9E6C8367EAB5C4831475E2081FFA656D57EADAC61",
            "B6DFAE6B40BDBC10D9805C4CC5867D804C8AA997B96445E16834D31017E3BFB6",
            "C6BC6D6FB2F77F03F178200D483FD20CE1A7FDE87E97629040CA953782AC002D",
            "D2FCDC13529A8586B0313F78E6499DCA5D946236007811139170DF7AED866870",
            "D8131A724817F4FDE969B52A924B44E984D94B0E2433A7AFBE056203F1BE8F41",
            "DBBF033763C839A026A1C7D6AAEFD88DDA72D7DD5617D40D2B5CACEDD2212156",
            "DE1A53940910DD8D69DEA1B1713E872A1CC16E02883187CA083FB3750474AA23",
            "E3750EE40504D7BF7FDDB9CEE16C9688590D2EE6F8FA01CC7A712C06F799727D",
            "EAA2DAED39BA5E506F99F5B446ECEF721684BA48E0783FFBD1A3EE748F8AC0CE",
            "F021CC33512D68ABD8AAE0A4C1AF1BEAE0D0637B11B32FC76691BD346CF34D4C",
            "F0FF32B7A58AF5AE394B893AB07280997138A4C7C1D298C114D7B9FADF2CB988",
            "F7689F4868E97B0B775AFF3939F7EBE0D5CFD950649829EF4C2019115381A603"
         ]
      },
      "ledger_hash" : "9DE6ECF7529B4DDB517A6D69D8994CE9B15C9DB1434E0203626180814A7E6873",
      "ledger_index" : 73469653,
      "status" : "success",
      "validated" : false
   }
}
$ rippled -q tx EAA2DAED39BA5E506F99F5B446ECEF721684BA48E0783FFBD1A3EE748F8AC0CE
{
   "result" : {
      "error" : "noNetwork",
      "error_code" : 17,
      "error_message" : "Not synced to the network.",
      "request" : {
         "api_version" : 1,
         "command" : "tx",
         "transaction" : "EAA2DAED39BA5E506F99F5B446ECEF721684BA48E0783FFBD1A3EE748F8AC0CE"
      },
      "status" : "error"
   }
}
```
As you can see, looking for the ledger by index doesn't work because the node isn't synced, but looking up by hash finds the ledger on disk, but doesn't yet know that it's validated (`"validated" : false`).

Unfortunately, my node keeps very little history, so once it synced, it very quickly finished loading all those old ledgers, but you can repeat the experiment on your node, since I'm not entirely sure what the behavior will be.

**Comment by 1143857824** [comment]
@ximinez Hi,ximinez
First of all, I will describe what I understand about what you said above to avoid misunderstanding.
1. That is to say, when the node is just started, it can only query the most recent block. Over time, the node will check forward the computation blocks. When the node verification is completed, the block from a long time ago can be queried .

Below is my method of querying blocks through nodes. As shown in the figure
![image](https://user-images.githubusercontent.com/57594569/182997104-cf7eae99-a170-416b-ac04-f4e24c186a9f.png)
**question:**
1. Our node was restarted two days ago, and data after the height "70144088" can be queried before the restart. After restarting, you can query the data after the height "72378219". Two days later, which is now, I still can only query the data after the height "72378219". There is no sign of forward synchronization, is it because we have not configured it correctly?
2. We found another case, when we removed these two configurations (online_delete=81920000 AND advisory_delete=0), after restarting the node, the node will indeed verify and synchronize the previous blocks. But when we add these two configurations, the node will not verify and synchronize previous blocks. I wonder if this is normal? And if we remove these two configs, will the node keep checking sync blocks forward until the height is "0" or will it only sync forward to the height "70144088" (the height when the node was first built)?

**Comment by ximinez** [comment]
1. I'm not sure why your node is not processing those older ledgers to be aware of them, but I'd bet they're still there. 
   * The hash of ledger [70144088 is `0566043611AC630A5B14161A214002CEA89434E2C1EFBF001F0D90F55F2F3C36`](https://livenet.xrpl.org/ledgers/70144088), so try something like `curl [...] '{"method", "ledger", "params": [{"ledger_hash": "0566043611AC630A5B14161A214002CEA89434E2C1EFBF001F0D90F55F2F3C36"}]'`.
   * The hash of ledger [72378218 is `D6E0E6378F93CAAC396D44C1ABB52F0DF3B287B284B237F702B4B2996720CE80`](https://livenet.xrpl.org/ledgers/72378218) so try something like `curl [...] '{"method", "ledger", "params": [{"ledger_hash": "D6E0E6378F93CAAC396D44C1ABB52F0DF3B287B284B237F702B4B2996720CE80"}]'`.
   * I expect that it will return unvalidated data for both ledgers, which indicates the data is on disk, but rippled is not looking back for it. That could be for several reasons, which I'm not going to go in to right now, because I _think_ most of them can be worked around by increasing your `[ledger_history]` value to be at least large enough to reach back to the ledgers you're trying to access - that would be on the order of 3500000. One significant exception is if your node does not have sufficient resources or connectivity to stay in sync with the network and backfill ledgers. Rippled will always prioritize staying up to date over finding or acquiring historical data.
2. More generally, though, your `online_delete` setting is 81,920,000, which is more than the total number of historical ledgers: 73,494,876. The way that online delete works is that nothing will be deleted until you have at least double the number of ledgers as the setting. That will probably mean you're covered for the next decade plus. You're effectively running a full history node, and it may make more sense to configure it explicitly to do that. Instructions are at: https://xrpl.org/configure-full-history.html

**Comment by 1143857824** [comment]
Hi ximinez 。
According to your statement, I tried to request, but I can't get the block data.
![image](https://user-images.githubusercontent.com/57594569/183367763-2c75bc79-62d5-435b-bb10-f7d2b6e9c637.png)

This means that there is no previous data after the node is restarted.  This is strange, but we do not intend to dwell on this issue.  Prepare to rebuild the new node.

We suspect that there is a problem with the directory mounted on the "docker container", which causes data loss after restart. I hope you can tell me where the data directory of the node is? Thank you.

The following is our current node building method. If there is any problem, please help me correct it， I'm very grateful. (we hope that the node can have data since the construction, but we don't need previous data.)
**docker-compose.yml**
    
    version: "3"
    
    networks:
      xrp-mainnet-01:
        ipam:
          driver: default
          config:
            - subnet: 192.168.76.0/24
    
    services:
      xrp-mainnet-01:
        container_name: wallet-xrp-mainnet-01
        hostname: wallet-xrp-mainnet-01
        image: image-wallet-xrp-daemon
        command: "/opt/ripple/bin/rippled --conf /etc/opt/ripple/rippled.cfg"
        ports:
          - "61076:5005"
        build:
          context: ./daemon
        environment:
        - TZ=Asia/Shanghai
        volumes:
          - /data/xrp-mainnet-01:/var/lib/rippled/
        networks:
          - xrp-mainnet-01
        tty: true
        mem_limit: 64g
        restart: always

**Dockerfile**

    FROM ubuntu:20.04

    RUN cd /root \
            && apt-get update && apt-get -y install apt-transport-https ca-certificates wget gnupg \
	    && wget -q -O - "https://repos.ripple.com/repos/api/gpg/key/public" | apt-key add - \
	    && echo "deb https://repos.ripple.com/repos/rippled-deb bionic stable" | tee -a /etc/apt/sources.list.d/ripple.list \
	    && apt -y update \
	    && apt -y install rippled \
	    && apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*

    COPY rippled.cfg /etc/opt/ripple/rippled.cfg

**rippled.cfg**
    
    [server]
    port_rpc_admin_local
    port_peer
    port_ws_admin_local
    
    [port_rpc_admin_local]
    port = 5005
    ip = 0.0.0.0
    admin = 0.0.0.0
    protocol = http
    
    [port_peer]
    port = 51235
    ip = 0.0.0.0
    #ip = ::
    protocol = peer
    
    [port_ws_admin_local]
    port = 6006
    ip = 0.0.0.0
    admin = 0.0.0.0
    protocol = ws
    
    [node_size]
    medium
    
    [node_db]
    type=NuDB
    path=/var/lib/rippled/db/nudb
    online_delete=20480000
    advisory_delete=0
    
    [database_path]
    /var/lib/rippled/db
    
    [debug_logfile]
    /var/log/rippled/debug.log
    
    [sntp_servers]
    time.windows.com
    time.apple.com
    time.nist.gov
    pool.ntp.org
    
    
    [validators_file]
    validators.txt
    
    [rpc_startup]
    { "command": "log_level", "severity": "warning" }
    
    [ssl_verify]
    1

**Comment by ximinez** [comment]
I'm not a Docker expert, and I can't really comment on anything related to that. You're data is stored based on the location you specify in the ripple.cfg file, in this case:
```
[node_db]
path=/var/lib/rippled/db/nudb

[database_path]
/var/lib/rippled/db

[debug_logfile]
/var/log/rippled/debug.log
```
Probably relative to the Docker container.

**Comment by 1143857824** [comment]
OK, thank you very much.
